// @shopify/polaris/dist/esm/components/IndexTable/components/Checkbox/Checkbox.js
import React$1, { memo, useContext, Fragment, useRef, useCallback, useEffect } from 'react';
import { useI18n } from '../../../../utilities/i18n/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { Checkbox as Checkbox$3 } from '../../../Checkbox/Checkbox.js';
import { setRootProperty as setRootProperty$1 } from '../../../../utilities/set-root-property.js';
import styles$1 from '../../IndexTable.scss.js';
import { useIndexValue } from '../../../../utilities/index-provider/hooks.js';
import { RowContext } from '../../../../utilities/index-table/context.js';
import styles from './Checkbox.scss.js';

const Checkbox = /*#__PURE__*/memo(function Checkbox() {
  const i18n = useI18n();
  const {
    resourceName,
    condensed
  } = useIndexValue();
  const {
    itemId,
    selected,
    onInteraction
  } = useContext(RowContext);
  const wrapperClassName = classNames(styles.Wrapper, condensed ? styles.condensed : styles.expanded);
  const Wrapper = condensed ? Fragment : CheckboxWrapper;
  return /*#__PURE__*/React$1.createElement(Wrapper, null, /*#__PURE__*/React$1.createElement("div", {
    className: styles.TableCellContentContainer
  }, /*#__PURE__*/React$1.createElement("div", {
    className: wrapperClassName,
    onClick: onInteraction,
    onKeyUp: onInteraction,
    onChange: stopPropagation
  }, /*#__PURE__*/React$1.createElement(Checkbox$3, {
    id: itemId,
    label: i18n.translate('Polaris.IndexTable.selectItem', {
      resourceName: resourceName.singular
    }),
    labelHidden: true,
    checked: selected
  }))));
});
function CheckboxWrapper({
  children
}) {
  const checkboxNode = useRef(null);
  const handleResize = useCallback(() => {
    if (!checkboxNode.current) return;
    const {
      width
    } = checkboxNode.current.getBoundingClientRect();
    setRootProperty$1('--p-checkbox-offset', `${width}px`, null);
  }, []);
  useEffect(() => {
    handleResize();
  }, [handleResize]);
  useEffect(() => {
    if (!checkboxNode.current) return;
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [handleResize]);
  const checkboxClassName = classNames(styles$1.TableCell, styles$1['TableCell-first']);
  return /*#__PURE__*/React$1.createElement("td", {
    className: checkboxClassName,
    ref: checkboxNode
  }, children);
}

function stopPropagation(event) {
  event.stopPropagation();
}

export { Checkbox, CheckboxWrapper };
