// @shopify/polaris/dist/esm/components/Tabs/components/List/List.js
import React$1 from 'react';
import { objectWithoutProperties as _objectWithoutProperties } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import styles from '../../Tabs.scss.js';
import { Item as Item$1 } from '../Item/Item.js';

function List({
  focusIndex,
  disclosureTabs,
  onClick = noop,
  onKeyPress = noop
}) {
  const tabs = disclosureTabs.map((_ref, index) => {
    let {
      id,
      content
    } = _ref,
        tabProps = _objectWithoutProperties(_ref, ["id", "content"]);

    return /*#__PURE__*/React$1.createElement(Item$1, Object.assign({}, tabProps, {
      key: id,
      id: id,
      focused: index === focusIndex,
      onClick: onClick.bind(null, id)
    }), content);
  });
  return /*#__PURE__*/React$1.createElement("ul", {
    className: styles.List,
    onKeyDown: handleKeyDown,
    onKeyUp: onKeyPress
  }, tabs);
}

function noop() {}

function handleKeyDown(event) {
  const {
    key
  } = event;

  if (key === 'ArrowLeft' || key === 'ArrowRight') {
    event.preventDefault();
    event.stopPropagation();
  }
}

export { List };
